jvb/src/test/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionTest.kt

```diff
@@ -1010,13 +1010,172 @@ class Vp9AdaptiveSourceProjectionTest {
         runLargeDropoutTest(generator, 0)
     }
 
+    private fun runSourceSuspensionTest(generator: Vp9PacketGenerator, targetIndex: Int) {
+        val diagnosticContext = DiagnosticContext()
+        diagnosticContext["test"] = Thread.currentThread().stackTrace[2].methodName
+        val initialState = RtpState(1, 10000, 1000000)
+        val context = Vp9AdaptiveSourceProjectionContext(
+            diagnosticContext,
+            payloadType,
+            initialState, logger
+        )
+        var expectedSeq = 10001
+        var expectedTs: Long = 1003000
+        var expectedPicId = 0
+        var expectedTl0PicIdx = 0
+        val targetSid = getSidFromIndex(targetIndex)
+        val targetTid = getTidFromIndex(targetIndex)
+
+        var packetInfo: PacketInfo
+        var packet: Vp9Packet
+
+        for (i in 0..999) {
+            packetInfo = generator.nextPacket()
+            packet = packetInfo.packetAs()
+            val accepted = context.accept(
+                packetInfo,
+                getIndex(0, packet.spatialLayerIndex, packet.temporalLayerIndex), targetIndex
+            )
+            if (packet.isStartOfFrame && packet.temporalLayerIndex == 0) {
+                expectedTl0PicIdx = applyTl0PicIdxDelta(expectedTl0PicIdx, 1)
+            }
+            val endOfPicture = packet.isEndOfPicture
+            if (packet.temporalLayerIndex <= targetTid &&
+                (
+                    packet.spatialLayerIndex == targetSid ||
+                        (packet.isUpperLevelReference && packet.spatialLayerIndex < targetSid)
+                    )
+            ) {
+                Assert.assertTrue(accepted)
+                context.rewriteRtp(packetInfo)
+                Assert.assertEquals(expectedSeq, packet.sequenceNumber)
+                Assert.assertEquals(expectedTs, packet.timestamp)
+                Assert.assertEquals(expectedPicId, packet.pictureId)
+                Assert.assertEquals(expectedTl0PicIdx, packet.TL0PICIDX)
+                expectedSeq = RtpUtils.applySequenceNumberDelta(expectedSeq, 1)
+            } else {
+                Assert.assertFalse(accepted)
+            }
+            if (endOfPicture) {
+                expectedTs = RtpUtils.applyTimestampDelta(expectedTs, 3000)
+                expectedPicId = applyExtendedPictureIdDelta(expectedPicId, 1)
+            }
+        }
+        var suspended = 64
+        while (suspended < 65536) {
```

Comment:
This loop is equivalent to `for(suspended = 64; suspended < 65536; suspended *= 2)`, right? I find the `for` version easier to read.

Reply:
Kotlin doesn't actually have C-style `for` loops, only `foreach`-style loops.

I can use a helper function of the type seen at https://stackoverflow.com/questions/66377331/how-do-you-iterate-a-for-loop-with-multiplication-or-division-kotlin .

Reply:
That's annoying. You could also do something like
```
for (i in 5..16) {
  val suspended = 2.0.pow(i).toInt()
}
```

I'm fine with either version, including the original.